go.mod

```diff
@@ -11,12 +11,40 @@ require (
 	github.com/prometheus/prometheus v2.5.0+incompatible
 	github.com/robfig/cron/v3 v3.0.0
 	github.com/spf13/pflag v1.0.3
-	golang.org/x/tools v0.0.0-20180917221912-90fa682c2a6e
-	k8s.io/api v0.0.0-20190620084959-7cf5895f2711
-	k8s.io/apimachinery v0.0.0-20190612205821-1799e75a0719
-	k8s.io/autoscaler v0.0.0-20190607113959-1b4f1855cb8e
-	k8s.io/client-go v0.0.0-20190620085101-78d2af792bab
-	k8s.io/klog v0.4.0
+	golang.org/x/tools v0.0.0-20190621195816-6e04913cbbac
+	k8s.io/api v0.0.0
+	k8s.io/apimachinery v0.0.0
+	k8s.io/autoscaler/vertical-pod-autoscaler v0.0.0-20191031123536-086999e99674
+	k8s.io/client-go v0.0.0
+	k8s.io/klog v1.0.0
 )
 
 go 1.13
+
+replace (
```

Comment:
Do we need to have these replace directives?

Reply:
tl;dr I wasn't able to find a way without them

https://github.com/kubernetes/autoscaler/pull/2486 has updated the outdated kubernetes dependencies to v1.16.2 and also added go modules with dependency on `k8s.io/kubernetes` [directly.](https://github.com/kubernetes/autoscaler/blob/3da8964ead179c471f6a2d17dde9b35b935df2ab/vertical-pod-autoscaler/go.mod#L15-L21)

Summarizing my understanding of discussion in https://github.com/kubernetes/kubernetes/issues/79384, you [should not](https://github.com/kubernetes/kubernetes/issues/79384#issuecomment-505725449) depend on `k8s.io/kubernetes`, but if you still do (or if you depend on [something that does](https://github.com/kubernetes/kubernetes/issues/79384#issuecomment-538035379)), you should pin the deps [accordingly](https://github.com/kubernetes/kubernetes/issues/79384#issuecomment-505627280). 

A script that I quite like that helps to figure out the exact versions for those replace directives https://github.com/kubernetes/kubernetes/issues/79384#issuecomment-521493597 was also mentioned in that thread. That said, I will support any idea on how to improve this but I couldn't figure out anything. Well, apart from trying to make VPA not depend on `k8s.io/kubernetes`, but that may be quite hard, if possible at all right now.

Reply:
This is what I see in VPA *gasp* - https://github.com/kubernetes/autoscaler/blob/3da8964ead179c471f6a2d17dde9b35b935df2ab/vertical-pod-autoscaler/go.mod#L338

Reply:
Depending on `kubernetes/kubernetes` itself is a huge red flag. We had previously worked towards removing that dependency in `kube-state-metrics` keeping in mind the multiple problems that could arise from it. I don't think we should incorporate changes unless we have these fixed in upstream. We have always strongly advocated for maintaining clean dependency trees and I think it would be best to ensure the status quo.

I am happy to dive deeper into this and also join in on. discussions with upstream projects about this.

Reply:
sounds good to me. I have done a quick check on what may be the reasons for depending on `k8s.io/kubernetes` in VPA and it looks like the largest part is in e2e tests depending on `k8s.io/kubernetes/tests/...` packages and one import in `pkg/utils/metrics/metrics.go` depending on `k8s.io/kubernetes/pkg/client/metrics/prometheus`.

Existence of `k8s.io/tests` package would make this quite simple but I am not aware of such repo. Do you have any ideas? Alternatively, maybe we can convince VPA maintainers to move e2e tests to different package and make `k8s.io/autoscaler/vertical-pod-autoscale` free of `k8.io/kubernetes` imports.

<details><summary>grep output</summary>
<p>

```
$ grep k8s.io/autoscaler/vertical-pod-autoscaler --include=*.go -nre '[[:blank:]]"k8s.io/kubernetes' | grep -v "$vendor/k8s.io/kubernetes"
./pkg/utils/metrics/metrics.go:28:	_ "k8s.io/kubernetes/pkg/client/metrics/prometheus" // for client-go metrics registration
./e2e/v1/updater.go:33:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1/updater.go:34:	framework_deployment "k8s.io/kubernetes/test/e2e/framework/deployment"
./e2e/v1/updater.go:35:	framework_job "k8s.io/kubernetes/test/e2e/framework/job"
./e2e/v1/updater.go:36:	framework_rs "k8s.io/kubernetes/test/e2e/framework/replicaset"
./e2e/v1/updater.go:37:	framework_ss "k8s.io/kubernetes/test/e2e/framework/statefulset"
./e2e/v1/updater.go:38:	testutils "k8s.io/kubernetes/test/utils"
./e2e/v1/autoscaling_utils.go:34:	api "k8s.io/kubernetes/pkg/apis/core"
./e2e/v1/autoscaling_utils.go:35:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1/autoscaling_utils.go:36:	"k8s.io/kubernetes/test/e2e/framework/replicaset"
./e2e/v1/autoscaling_utils.go:37:	e2eservice "k8s.io/kubernetes/test/e2e/framework/service"
./e2e/v1/autoscaling_utils.go:38:	testutils "k8s.io/kubernetes/test/utils"
./e2e/v1/autoscaling_utils.go:42:	imageutils "k8s.io/kubernetes/test/utils/image"
./e2e/v1/e2e_test.go:33:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1/e2e_test.go:34:	"k8s.io/kubernetes/test/e2e/framework/config"
./e2e/v1/e2e_test.go:35:	"k8s.io/kubernetes/test/e2e/framework/testfiles"
./e2e/v1/e2e_test.go:36:	"k8s.io/kubernetes/test/e2e/framework/viperconfig"
./e2e/v1/e2e_test.go:37:	"k8s.io/kubernetes/test/e2e/generated"
./e2e/v1/e2e_test.go:38:	"k8s.io/kubernetes/test/utils/image"
./e2e/v1/e2e_test.go:41:	_ "k8s.io/kubernetes/test/e2e/apimachinery"
./e2e/v1/e2e_test.go:42:	_ "k8s.io/kubernetes/test/e2e/apps"
./e2e/v1/e2e_test.go:43:	_ "k8s.io/kubernetes/test/e2e/auth"
./e2e/v1/e2e_test.go:44:	_ "k8s.io/kubernetes/test/e2e/autoscaling"
./e2e/v1/e2e_test.go:45:	_ "k8s.io/kubernetes/test/e2e/cloud"
./e2e/v1/e2e_test.go:46:	_ "k8s.io/kubernetes/test/e2e/common"
./e2e/v1/e2e_test.go:47:	_ "k8s.io/kubernetes/test/e2e/instrumentation"
./e2e/v1/e2e_test.go:48:	_ "k8s.io/kubernetes/test/e2e/kubectl"
./e2e/v1/e2e_test.go:49:	_ "k8s.io/kubernetes/test/e2e/lifecycle"
./e2e/v1/e2e_test.go:50:	_ "k8s.io/kubernetes/test/e2e/lifecycle/bootstrap"
./e2e/v1/e2e_test.go:51:	_ "k8s.io/kubernetes/test/e2e/network"
./e2e/v1/e2e_test.go:52:	_ "k8s.io/kubernetes/test/e2e/node"
./e2e/v1/e2e_test.go:53:	_ "k8s.io/kubernetes/test/e2e/scalability"
./e2e/v1/e2e_test.go:54:	_ "k8s.io/kubernetes/test/e2e/scheduling"
./e2e/v1/e2e_test.go:55:	_ "k8s.io/kubernetes/test/e2e/servicecatalog"
./e2e/v1/e2e_test.go:56:	_ "k8s.io/kubernetes/test/e2e/storage"
./e2e/v1/e2e_test.go:57:	_ "k8s.io/kubernetes/test/e2e/storage/external"
./e2e/v1/e2e_test.go:58:	_ "k8s.io/kubernetes/test/e2e/ui"
./e2e/v1/e2e_test.go:59:	_ "k8s.io/kubernetes/test/e2e/windows"
./e2e/v1/recommender.go:32:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1/recommender.go:33:	framework_deployment "k8s.io/kubernetes/test/e2e/framework/deployment"
./e2e/v1/common.go:37:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1/common.go:38:	framework_deployment "k8s.io/kubernetes/test/e2e/framework/deployment"
./e2e/v1/actuation.go:30:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1/actuation.go:31:	framework_deployment "k8s.io/kubernetes/test/e2e/framework/deployment"
./e2e/v1/e2e.go:34:	commontest "k8s.io/kubernetes/test/e2e/common"
./e2e/v1/e2e.go:35:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1/e2e.go:36:	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
./e2e/v1/e2e.go:42:	_ "k8s.io/kubernetes/test/e2e/framework/providers/aws"
./e2e/v1/e2e.go:43:	_ "k8s.io/kubernetes/test/e2e/framework/providers/azure"
./e2e/v1/e2e.go:44:	_ "k8s.io/kubernetes/test/e2e/framework/providers/gce"
./e2e/v1/e2e.go:45:	_ "k8s.io/kubernetes/test/e2e/framework/providers/kubemark"
./e2e/v1/e2e.go:46:	_ "k8s.io/kubernetes/test/e2e/framework/providers/openstack"
./e2e/v1/e2e.go:47:	_ "k8s.io/kubernetes/test/e2e/framework/providers/vsphere"
./e2e/v1/admission_controller.go:27:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1/admission_controller.go:28:	framework_deployment "k8s.io/kubernetes/test/e2e/framework/deployment"
./e2e/v1/full_vpa.go:29:	e2e_common "k8s.io/kubernetes/test/e2e/common"
./e2e/v1/full_vpa.go:30:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1beta2/updater.go:33:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1beta2/updater.go:34:	framework_deployment "k8s.io/kubernetes/test/e2e/framework/deployment"
./e2e/v1beta2/updater.go:35:	framework_job "k8s.io/kubernetes/test/e2e/framework/job"
./e2e/v1beta2/updater.go:36:	framework_rs "k8s.io/kubernetes/test/e2e/framework/replicaset"
./e2e/v1beta2/updater.go:37:	framework_ss "k8s.io/kubernetes/test/e2e/framework/statefulset"
./e2e/v1beta2/updater.go:38:	testutils "k8s.io/kubernetes/test/utils"
./e2e/v1beta2/autoscaling_utils.go:34:	api "k8s.io/kubernetes/pkg/apis/core"
./e2e/v1beta2/autoscaling_utils.go:35:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1beta2/autoscaling_utils.go:36:	"k8s.io/kubernetes/test/e2e/framework/replicaset"
./e2e/v1beta2/autoscaling_utils.go:37:	e2eservice "k8s.io/kubernetes/test/e2e/framework/service"
./e2e/v1beta2/autoscaling_utils.go:38:	testutils "k8s.io/kubernetes/test/utils"
./e2e/v1beta2/autoscaling_utils.go:42:	imageutils "k8s.io/kubernetes/test/utils/image"
./e2e/v1beta2/e2e_test.go:33:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1beta2/e2e_test.go:34:	"k8s.io/kubernetes/test/e2e/framework/config"
./e2e/v1beta2/e2e_test.go:35:	"k8s.io/kubernetes/test/e2e/framework/testfiles"
./e2e/v1beta2/e2e_test.go:36:	"k8s.io/kubernetes/test/e2e/framework/viperconfig"
./e2e/v1beta2/e2e_test.go:37:	"k8s.io/kubernetes/test/e2e/generated"
./e2e/v1beta2/e2e_test.go:38:	"k8s.io/kubernetes/test/utils/image"
./e2e/v1beta2/e2e_test.go:41:	_ "k8s.io/kubernetes/test/e2e/apimachinery"
./e2e/v1beta2/e2e_test.go:42:	_ "k8s.io/kubernetes/test/e2e/apps"
./e2e/v1beta2/e2e_test.go:43:	_ "k8s.io/kubernetes/test/e2e/auth"
./e2e/v1beta2/e2e_test.go:44:	_ "k8s.io/kubernetes/test/e2e/autoscaling"
./e2e/v1beta2/e2e_test.go:45:	_ "k8s.io/kubernetes/test/e2e/cloud"
./e2e/v1beta2/e2e_test.go:46:	_ "k8s.io/kubernetes/test/e2e/common"
./e2e/v1beta2/e2e_test.go:47:	_ "k8s.io/kubernetes/test/e2e/instrumentation"
./e2e/v1beta2/e2e_test.go:48:	_ "k8s.io/kubernetes/test/e2e/kubectl"
./e2e/v1beta2/e2e_test.go:49:	_ "k8s.io/kubernetes/test/e2e/lifecycle"
./e2e/v1beta2/e2e_test.go:50:	_ "k8s.io/kubernetes/test/e2e/lifecycle/bootstrap"
./e2e/v1beta2/e2e_test.go:51:	_ "k8s.io/kubernetes/test/e2e/network"
./e2e/v1beta2/e2e_test.go:52:	_ "k8s.io/kubernetes/test/e2e/node"
./e2e/v1beta2/e2e_test.go:53:	_ "k8s.io/kubernetes/test/e2e/scalability"
./e2e/v1beta2/e2e_test.go:54:	_ "k8s.io/kubernetes/test/e2e/scheduling"
./e2e/v1beta2/e2e_test.go:55:	_ "k8s.io/kubernetes/test/e2e/servicecatalog"
./e2e/v1beta2/e2e_test.go:56:	_ "k8s.io/kubernetes/test/e2e/storage"
./e2e/v1beta2/e2e_test.go:57:	_ "k8s.io/kubernetes/test/e2e/storage/external"
./e2e/v1beta2/e2e_test.go:58:	_ "k8s.io/kubernetes/test/e2e/ui"
./e2e/v1beta2/e2e_test.go:59:	_ "k8s.io/kubernetes/test/e2e/windows"
./e2e/v1beta2/recommender.go:32:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1beta2/recommender.go:33:	framework_deployment "k8s.io/kubernetes/test/e2e/framework/deployment"
./e2e/v1beta2/common.go:37:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1beta2/common.go:38:	framework_deployment "k8s.io/kubernetes/test/e2e/framework/deployment"
./e2e/v1beta2/actuation.go:30:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1beta2/actuation.go:31:	framework_deployment "k8s.io/kubernetes/test/e2e/framework/deployment"
./e2e/v1beta2/e2e.go:34:	commontest "k8s.io/kubernetes/test/e2e/common"
./e2e/v1beta2/e2e.go:35:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1beta2/e2e.go:36:	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
./e2e/v1beta2/e2e.go:42:	_ "k8s.io/kubernetes/test/e2e/framework/providers/aws"
./e2e/v1beta2/e2e.go:43:	_ "k8s.io/kubernetes/test/e2e/framework/providers/azure"
./e2e/v1beta2/e2e.go:44:	_ "k8s.io/kubernetes/test/e2e/framework/providers/gce"
./e2e/v1beta2/e2e.go:45:	_ "k8s.io/kubernetes/test/e2e/framework/providers/kubemark"
./e2e/v1beta2/e2e.go:46:	_ "k8s.io/kubernetes/test/e2e/framework/providers/openstack"
./e2e/v1beta2/e2e.go:47:	_ "k8s.io/kubernetes/test/e2e/framework/providers/vsphere"
./e2e/v1beta2/admission_controller.go:27:	"k8s.io/kubernetes/test/e2e/framework"
./e2e/v1beta2/admission_controller.go:28:	framework_deployment "k8s.io/kubernetes/test/e2e/framework/deployment"
./e2e/v1beta2/full_vpa.go:29:	e2e_common "k8s.io/kubernetes/test/e2e/common"
./e2e/v1beta2/full_vpa.go:30:	"k8s.io/kubernetes/test/e2e/framework"
./vendor/k8s.io/kubelet/config/v1beta1/types.go:655:	// "k8s.io/kubernetes/pkg/features/kube_features.go".
./vendor/k8s.io/cloud-provider/service/helpers/helper.go:28:This file is duplicated from "k8s.io/kubernetes/pkg/api/v1/service/util.go"
./vendor/k8s.io/cli-runtime/pkg/printers/sourcechecker.go:29:		"k8s.io/kubernetes/pkg/apis/"
```

</p>
</details>

Reply:
@tariq1890, ptal https://github.com/kubernetes/autoscaler/pull/2531

Reply:
deps updated to not contain any replace directives and/or dependency on `k8s.io/kubernetes` https://github.com/kubernetes/kube-state-metrics/pull/949/commits/4ffe41f11f1c4fb3eb2653bd4e71dfdc4372f52b

ptal @tariq1890

Reply:
Wohoo! Thanks so much for looking into this @wozniakjan! Much appreciated!